lambdas/check-existing-identity/src/main/java/uk/gov/di/ipv/core/checkexistingidentity/CheckExistingIdentityHandler.java

```diff
@@ -331,22 +333,78 @@ private JourneyResponse getJourneyResponse(
             var vcs =
                     evcsService.getVerifiableCredentialsByState(
-                            userId, evcsAccessToken, CURRENT, PENDING_RETURN);
-            var pendingReturnVcs = vcs.get(PENDING_RETURN);
-            // use pending return vcs to determine identity if available
-            if (!isNullOrEmpty(pendingReturnVcs)) {
-                return new VerifiableCredentialBundle(pendingReturnVcs, true, true);
+                            userId, evcsAccessToken, CURRENT, PENDING_RETURN, PENDING);
+
+            if (configService.enabled(EVCS_WRITE_ENABLED)) {
+                logIdentityMismatches(tacticalVcs, vcs);
             }
-            var currentVcs = vcs.get(CURRENT);
-            if (!isNullOrEmpty(currentVcs)) {
-                return new VerifiableCredentialBundle(currentVcs, true, false);
+            if (configService.enabled(EVCS_READ_ENABLED)) {
+                var pendingReturnVcs = vcs.get(PENDING_RETURN);
+                // use pending return vcs to determine identity if available
+                if (!isNullOrEmpty(pendingReturnVcs)) {
+                    return new VerifiableCredentialBundle(pendingReturnVcs, true, true);
+                }
+                var currentVcs = vcs.get(CURRENT);
+                if (!isNullOrEmpty(currentVcs)) {
+                    return new VerifiableCredentialBundle(currentVcs, true, false);
+                }
             }
         }
-        return new VerifiableCredentialBundle(
-                verifiableCredentialService.getVcs(userId), false, false);
+        return new VerifiableCredentialBundle(tacticalVcs, false, false);
+    }
+
+    @ExcludeFromGeneratedCoverageReport
+    private void logIdentityMismatches(
+            List<VerifiableCredential> tacticalVcs,
+            Map<EvcsVCState, List<VerifiableCredential>> evcsEvcs) {
+
+        // if we have pending vcs just check those
+        var evcsToCheck = evcsEvcs.get(PENDING);
+        if (isNullOrEmpty(evcsToCheck)) {
+            evcsToCheck = evcsEvcs.get(CURRENT);
+        }
+        // filter out unmigrated vcs from the tactical vcs
+        tacticalVcs = tacticalVcs.stream().filter(vc -> vc.getMigrated() != null).toList();
+
+        // check all the tactical vcs are in the selected evcs vcs
+        if (hasIdentityMismatch(tacticalVcs, evcsToCheck)) {
+            LOGGER.warn(
+                    "Failed to find corresponding evcs credential for migrated tactical credential");
+        }
+
+        // check all the evcs vcs are in the tactical store
+        var listOfAllEvcsVcs =
+                evcsEvcs.values().stream().flatMap(List<VerifiableCredential>::stream).toList();
+        if (hasIdentityMismatch(listOfAllEvcsVcs, tacticalVcs)) {
+            LOGGER.warn("Failed to find corresponding tactical credential for evcs credential");
+        }
+    }
+
+    @ExcludeFromGeneratedCoverageReport
```

Comment:
I was struggling to get the required test coverage in sonar cloud - is there a way to test private methods in java?

Reply:
Sorry forgot to reply to this, _in general_ either:
- The private method is an implementation detail, and should be be exercised via the public methods of the class
- It's not an implementation detail, so pull it out into a helper (or make it public) and test it that way

This is a bit of a weird case, because it only does logging, which we're not well set up to test. It's still possible (you can mock out the logger if you use reflection, or pull out a helper that gives you the 'result' to log, with the logging itself separate), but the effort might not outweigh the cost.